### PR TITLE
Refactor user import, support user data in import and fix profile data bug

### PR DESCRIPTION
**Pull request by sravfeyn**
## Product Description

This is a Refactor https://dimagi-dev.atlassian.net/browse/USH-3613 and a bug fix https://dimagi-dev.atlassian.net/browse/USH-2762

## Technical Summary

https://dimagi-dev.atlassian.net/browse/USH-3613

Broken up the long function methods into classes to improve reusability and (hopefully) better readability.

## Feature Flag

NA

### Safety story

I have tested the import locally and there is a good amount of test cases for the code that's refactored.

### Automated test coverage

`corehq.apps.user_importer.tests`

### QA Plan

Perhaps not required since the basic import works and there is enough test coverage 

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by sravfeyn** [comment]
@esoergel

**Review by esoergel — COMMENTED** [review]
Things are looking much better overall - I had difficulty figuring out what changed since it's so much code all at once, but I had some hopefully-useful comments.  Any idea how good test coverage is on this?

File: corehq/apps/user_importer/importer.py [hunk 1]
```diff
@@ -697,13 +480,410 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
+        elif self.column_values["send_confirmation_sms"]:
+            # Set a dummy password to pass the validation, similar to GUI user creation
+            string_set = string.ascii_uppercase + string.digits + string.ascii_lowercase
+            password = ''.join(random.choices(string_set, k=10))
+        else:
+            password = None
+        self.column_values['password'] = password
+        self.status_row['row']['password'] = password
+        if self.column_values['user_id'] and is_password(password):
+            self.status_row['row']['password'] = 'REDACTED'
```
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/user_importer/importer.py:509]: > Why does this check if user_id? Shouldn't it always redact passwords, also for new users?

I was using the older logic. I guess, we could redact. I refactored without changing the logic, I will update.

> It looks like the old version of the code doesn't set password at all if the user ID isn't present.

Right, we are still doing the [same](https://github.com/dimagi/commcare-hq/pull/33890/files#diff-3875199418c65c97fec9e657efd64668638b4bb447d3a24a1ad3dd42e14064c4R567), no?

File: corehq/apps/user_importer/importer.py [hunk 2]
```diff
@@ -697,13 +480,410 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
+        else:
+            values["is_account_confirmed"] = values["is_account_confirmed"]
+
+        self.column_values.update(values)
+        if not self._parse_username():
+            return False
+        self._parse_password()
+        return True
+
+    def _set_user_importer(self):
```
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/user_importer/importer.py:544]: That's a good point, updated.

File: corehq/apps/user_importer/importer.py [hunk 3]
```diff
@@ -697,13 +480,410 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
+            self.validate_row()
+            role = self.row.get('role')
+            status = self.row.get('status')
+            location_codes = self.row.get('location_code', [])
+            location_codes = format_location_codes(location_codes)
+            remove = spec_value_to_boolean_or_none(self.row, 'remove')
+
+            self.process_row(role, status, location_codes, remove)
+        except UserUploadError as e:
+            self.status_row['flag'] = str(e)
```
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/user_importer/importer.py:679]: Yes, we can move it. (It's still [here](https://github.com/dimagi/commcare-hq/pull/33890/files#diff-3875199418c65c97fec9e657efd64668638b4bb447d3a24a1ad3dd42e14064c4R484) in commcare_user version as well).

File: corehq/apps/user_importer/importer.py [hunk 4]
```diff
@@ -697,13 +480,410 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
 
-    return ret
+    @memoized
+    def domain_info(self, domain):
+        return DomainInfo(self, domain, is_web_upload=self.is_web_upload)
+
+    def run(self):
+        ret = {"errors": [], "rows": []}
+        current = 0
+        for row in self.user_specs:
```
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/user_importer/importer.py:779]: Of course!

File: corehq/apps/user_importer/importer.py [hunk 5]
```diff
@@ -697,13 +480,410 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
...
+
+
+class DomainInfo:
+
+    def __init__(self, importer, domain, is_web_upload):
+        self.importer = importer
+        self.domain = domain
+        self.is_web_upload = is_web_upload
+
+    @property
```
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/user_importer/importer.py:814]: That's right, updated it.

File: corehq/apps/user_importer/importer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/user_importer/importer.py:507]: There is this test that's failing `corehq.apps.user_importer.tests.test_importer:TestMobileUserBulkUpload.test_password_is_not_string` when I skip the user_id check. I am not sure why that's the way it is. I will rather leave the logic as is.

[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/user_importer/importer.py:507]: > It looks like the old version of the code doesn't set password at all if the user ID isn't present.

> Right, we are still doing the same, no?

I meant line 505 here, where password is added to the status row - just want to be sure there's no situation where we might store the password in plaintext.
```python
        self.status_row['row']['password'] = password
```
Would it be possible to just delete that line or initialize it to `""` or something?

File: corehq/apps/user_importer/tests/test_importer.py [hunk 6]
```diff
@@ -695,6 +695,28 @@ def test_user_data_profile_blank(self):
             PROFILE_SLUG: self.profile.id,
         })
 
+    def test_required_field_optional_if_profile_set(self):
```
Comment (esoergel) [inline corehq/apps/user_importer/tests/test_importer.py:698]: :star:

**Review by esoergel — APPROVED** [review]
apologies for the delay in getting back to the review - looks like this is about done

**Comment by esoergel** [comment]
Oh whoops, one note - this line in the description is no longer accurate:

> This is a Refactor without any product changes https://dimagi-dev.atlassian.net/browse/USH-3613

This PR now fixes that user data profile bug

File: corehq/apps/user_importer/importer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/user_importer/importer.py:507]: It looks like the status_row data is used only for communicating back to the uploading user when the password is not valid. If it is a valid password and it's set then it gets REDACTED. That's my understanding of the old logic. But of course there should be no problem to set that to empty regardless of whether the password is valid or not. I have updated it as such.

**Comment by sravfeyn** [comment]
@esoergel I have updated the description and also merged in the other PR https://github.com/dimagi/commcare-hq/pull/33930.

**Review by esoergel — APPROVED** [review]

**Comment by gherceg** [comment]
@sravfeyn is it fair to say you can now remove the `HELPME` comment in the `create_or_update_commcare_users_and_groups` method? I imagine it no longer triggers an 'E' or 'F' classification in static analysis.

**Pull request merged**